lib/instrumentation/transaction.js

```diff
@@ -218,3 +223,72 @@ Transaction.prototype.end = function (result, endTime) {
   this._agent._instrumentation.addEndedTransaction(this)
   this._agent.logger.debug('ended transaction %o', { trans: this.id, parent: this.parentId, trace: this.traceId, type: this.type, result: this.result, name: this.name })
 }
+
+Transaction.prototype._captureBreakdown = function (span) {
+  const metrics = this._agent._metrics
+
+  // Record span data
+  captureBreakdown(this, {
+    transaction: transactionBreakdownDetails(this),
+    span: spanBreakdownDetails(span)
+  }, span._timer.selfTime)
+
+  // Record transaction data
+  if (span instanceof Transaction) {
+    const { duration } = this._timer
+    const labels = flattenBreakdown({
+      transaction: transactionBreakdownDetails(this)
+    })
+
+    metrics.getOrCreateCounter('transaction.duration.count', labels).inc()
+    metrics.getOrCreateCounter('transaction.duration.sum.us', labels).inc(duration)
+    metrics.getOrCreateCounter('transaction.breakdown.count', labels).inc(this.sampled ? 1 : 0)
+
+    this._breakdownTimings.forEach(({ labels, time, count }) => {
+      labels = flattenBreakdown(labels)
+      metrics.getOrCreateCounter('span.self_time.count', labels).inc(count)
+      metrics.getOrCreateCounter('span.self_time.sum.us', labels).inc(time)
+    })
+  }
+}
+
+function transactionBreakdownDetails ({ name, type } = {}) {
+  return {
+    name,
+    type
+  }
+}
+
+function spanBreakdownDetails (span) {
+  if (span instanceof Transaction) {
+    return {
+      type: 'app'
+    }
+  }
+
+  const [ type, subtype ] = span.type.split('.')
+  return {
+    type,
+    subtype
+  }
+}
+
+function captureBreakdown (transaction, labels, time) {
+  const build = () => ({ labels, count: 0, time: 0 })
+  const counter = transaction._breakdownTimings.ensure(labels, build)
+  counter.time += time
+  counter.count++
+}
+
+function flattenBreakdown (source, target = {}, prefix = '') {
+  for (const [ key, value ] of entries(source)) {
+    if (typeof value === 'undefined') continue
+    if (typeof value === 'object') {
+      flattenBreakdown(source[key], target, `${prefix}${key}::`)
```

Comment:
is there a technical reason to use `::` over `.`?
(sorry if it's a dumb question, I'm asking for my own education)

Reply:
Less likely to be in an object key on purpose. 🤷‍♂

Reply:
Thought that might be it, but I also thought we wouldn't ever be including objects in the map. Anyway, I was just curious - thanks!